### PR TITLE
changefeedccl: enable frontier quantization

### LINKED ISSUE #125723: changefeedccl: take advantage of quantized frontiers
The DR time utilizes quantized frontiers (see https://github.com/cockroachdb/cockroach/pull/120631) to reduce the granularity of spans during checkpointing, allowing for smaller and less lossy checkpoints at the expense of potentially emitting more duplicates for some spans. CDC should consider adopting a similar strategy when checkpointing.

Jira issue: CRDB-39571

### PR DESCRIPTION
**Pull request by wenyihu6**
Previously, all checkpoint timestamps were forwarded precisely to the resolved
timestamp. To improve merging of adjacent frontier spans, this patch introduces
a frontier quantization mode. In this mode, resolved timestamps from rangefeeds
are rounded down to the nearest multiple of the quantization granularity.
Rounding down ensures no progress is missed and that we consistently track the
lowest possible resolved timestamp.

Resolves: https://github.com/cockroachdb/cockroach/issues/125723
Release note: none

### CODE REVIEW

File: pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go [hunk 1]
```diff
@@ -128,7 +130,12 @@ func (p *rangefeed) addEventsToBuffer(ctx context.Context) error {
 				}
 				stop()
 			case *kvpb.RangeFeedCheckpoint:
-				if !t.ResolvedTS.IsEmpty() && t.ResolvedTS.Less(p.cfg.Frontier) {
+				ev := e.RangeFeedEvent
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:133]: while this is a pretty simple implementation, would you mind investigating how much work it would be to switch to using the higher-level rangefeed client implementation that LDR uses? it already includes an implementation of quantization and more things that may prove useful. see https://github.com/asg0451/cockroach/blob/ed5ed37ea91b04fba061084ba73fe9225c018288/pkg/crosscluster/producer/event_stream.go#L147 &/ talk to me about it

Comment (asg0451) [reply pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:133]: if it's too much work / too incompatible an api, then maybe dont. however i think it would be beneficial if it's low to moderate effort

Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:133]: I will look into the refactoring. Can we try merging this PR anyways since this is pretty straightforward and already up? I will look into the refactoring as part of the follow up work.

Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:133]: I put up https://github.com/cockroachdb/cockroach/issues/138346 and assigned myself.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go [hunk 2]
```diff
@@ -343,8 +350,56 @@ func TestKVFeed(t *testing.T) {
...
+			events: []kvpb.RangeFeedEvent{
+				checkpointEvent(tableSpan(codec, 42), ts(2)),
+				checkpointEvent(tableSpan(codec, 42), hlc.Timestamp{WallTime: ts(2).WallTime, Logical: 1}),
+				checkpointEvent(tableSpan(codec, 42), hlc.Timestamp{WallTime: ts(3).WallTime + 5, Logical: 1}),
+				checkpointEvent(tableSpan(codec, 42), hlc.Timestamp{WallTime: ts(4).WallTime, Logical: 3}),
+			},
+			expEvents: []kvpb.RangeFeedEvent{
+				checkpointEvent(tableSpan(codec, 42), hlc.Timestamp{WallTime: ts(2).WallTime, Logical: 0}),
+				checkpointEvent(tableSpan(codec, 42), hlc.Timestamp{WallTime: ts(2).WallTime, Logical: 0}),
+				checkpointEvent(tableSpan(codec, 42), hlc.Timestamp{WallTime: ts(3).WallTime, Logical: 0}),
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:375]: for my own understanding, why is this 3? seems like it's not quantized to 10s like you specified. shouldnt it be ts(0) or ts(10) and nothing in between?

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go [hunk 3]
```diff
@@ -645,3 +700,151 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
...
+		ts.Logical = 0
+		ts.WallTime -= ts.WallTime % int64(quantize)
+		return ts
+	}
+
+	makeCheckpointEvent := func(key []byte, endKey []byte, ts int) *kvpb.RangeFeedEvent {
+		return &kvpb.RangeFeedEvent{
+			Checkpoint: &kvpb.RangeFeedCheckpoint{
+				Span:       roachpb.Span{Key: key, EndKey: endKey},
+				ResolvedTS: hlc.Timestamp{WallTime: int64(ts)},
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:718]: maybe add a logical time to this so you can test that it gets zero'd

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go [hunk 4]
```diff
@@ -645,3 +700,151 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
...
+			const quantize = time.Duration(10)
+			for _, e := range tc.events {
+				if tc.withFrontierQuantization {
+					e.Checkpoint.ResolvedTS = mockQuantization(e.Checkpoint.ResolvedTS, quantize)
+				}
+				_, err := frontier.Forward(e.Checkpoint.Span, e.Checkpoint.ResolvedTS)
+				require.NoError(t, err)
+			}
+			frontier.Entries(func(sp roachpb.Span, ts hlc.Timestamp) span.OpResult {
+				t.Logf("span: %v, ts: %v\n", sp, ts)
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:789]: nit rm

File: pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go (hunk 1 above)
Comment (asg0451) [reply pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:133]: thanks. lmk how it goes

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go [hunk 5]
```diff
@@ -645,3 +700,151 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
...
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(0)},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			frontier, err := span.MakeFrontier(tc.spans...)
+			require.NoError(t, err)
+			const quantize = time.Duration(10)
+			for _, e := range tc.events {
+				if tc.withFrontierQuantization {
+					e.Checkpoint.ResolvedTS = mockQuantization(e.Checkpoint.ResolvedTS, quantize)
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:783]: i'm not really sure what this test is testing. just your mockQuantization function? can you make sure it tests the actual implementation?

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go [hunk 6]
```diff
@@ -645,3 +700,151 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
...
+	}
+
+	makeCheckpointEvent := func(key []byte, endKey []byte, ts int, logical int32) *kvpb.RangeFeedEvent {
+		return &kvpb.RangeFeedEvent{
+			Checkpoint: &kvpb.RangeFeedCheckpoint{
+				Span:       roachpb.Span{Key: key, EndKey: endKey},
+				ResolvedTS: hlc.Timestamp{WallTime: int64(ts), Logical: logical},
+			},
+		}
+	}
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:815]: nit these can be pulled out between these two tests

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go [hunk 7]
```diff
@@ -645,3 +700,151 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
...
+	for i, s := range kvserverrangefeed.GenerateRandomizedSpans(rng, numOfSpans) {
+		ts := kvserverrangefeed.GenerateRandomizedTs(rng, time.Minute.Nanoseconds())
+		events[i] = makeCheckpointEvent(s.Key, s.EndKey, int(ts.WallTime), int32(rng.Intn(100)))
+	}
+
+	const quantize = time.Duration(10)
+	quantizedFrontier, err := span.MakeFrontier([]roachpb.Span{{Key: keys.MinKey, EndKey: keys.MaxKey}}...)
+	require.NoError(t, err)
+	for _, e := range events {
+		quantizedTs := mockQuantization(e.Checkpoint.ResolvedTS, quantize)
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:829]: ditto on the "what does this test" question

**Review by asg0451 — COMMENTED** [review]
looks good but i have some questions about the tests.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 2 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go]: It's related to the weirdness using the helper function ts - https://github.com/cockroachdb/cockroach/blob/cd1cba87ff145928bd18967d6d79b6489d870ae0/pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go#L53-L55. I refactored the test to avoid using this helper function.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 4 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:783]: I thought it might be worth leaving it in since it was helpful for debugging test failures wdyt?

File: pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go [hunk 8]
```diff
@@ -128,7 +139,11 @@ func (p *rangefeed) addEventsToBuffer(ctx context.Context) error {
...
+				if quantize := p.cfg.WithFrontierQuantize; quantize != 0 {
```
Comment (dt) [inline pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:143]: I don't know all the cdc code off the top of my head but one mistake I made and then spent some time tracking down in PCR/LDR is that we should only round down if doing so results in a ts that is still >= initial scan ts since otherwise we can prevent the initial scan from ever being considered complete.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 5 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go]: Good idea, I refactored the actual quantization to a helper function and made tests call into the helper function instead.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 7 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go]: I added more comments for these tests.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 3 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go]: Done.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 6 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go]: Done.

File: pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go (hunk 8 above)
Comment (dt) [reply pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:143]: we sync'ed up offline: this is only an issue in PCR because we quantize a second time when our coordinator is combining progress updates from each processor into the merged frontier and those progress updates are used both to make progress during initial scan and during steady state, which is why they need to be careful about moving a ts down in the former case. This diff here is just the RF case, not initial scans, so this doesn't apply.

File: pkg/ccl/changefeedccl/changefeedbase/settings.go [hunk 9]
```diff
@@ -335,3 +335,10 @@ var DefaultLaggingRangesThreshold = 3 * time.Minute
 // DefaultLaggingRangesPollingInterval is the default polling rate at which
 // lagging ranges are checked and metrics are updated.
 var DefaultLaggingRangesPollingInterval = 1 * time.Minute
+
+var Quantize = settings.RegisterDurationSettingWithExplicitUnit(
+	settings.ApplicationLevel,
+	"changefeed.frontier.timestamp_granularity",
```
Comment (andyyang890) [inline pkg/ccl/changefeedccl/changefeedbase/settings.go:341]: I'd prefer `changefeed.resolved_timestamp.granularity` since ultimately this will be surfaced to the user via quantized resolved timestamps. WDYT?

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go [hunk 10]
```diff
@@ -173,8 +177,11 @@ func TestKVFeed(t *testing.T) {
 		// Assert that number of events emitted from the kvfeed matches what we
 		// specified in the testcase.
 		testG.GoCtx(func(ctx context.Context) error {
-			for events := 0; events < tc.expEvents; events++ {
-				_, err := buf.Get(ctx)
+			for eventIdx := 0; eventIdx < tc.expEventsCount; eventIdx++ {
+				e, err := buf.Get(ctx)
+				if tc.expEvents != nil {
```
Comment (andyyang890) [inline pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:182]: nit: this if branch should go under the `assert.NoError`

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go [hunk 11]
```diff
@@ -645,3 +700,137 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
...
+		_, err := quantizedFrontier.Forward(e.Checkpoint.Span, quantizedTs)
+		require.NoError(t, err)
+	}
+	quantizedEntries := quantizedFrontier.Len()
+	quantizedHW := quantizedFrontier.Frontier()
+
+	frontier, err := span.MakeFrontier([]roachpb.Span{{Key: keys.MinKey, EndKey: keys.MaxKey}}...)
+	require.NoError(t, err)
+	for _, e := range events {
+		_, err := frontier.Forward(e.Checkpoint.Span, e.Checkpoint.ResolvedTS)
```
Comment (andyyang890) [inline pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:825]: nit: create both frontiers at the same time and have a single loop where you forward the events to both frontiers

**Review by andyyang890 — COMMENTED** [review]
Nice work! Left some comments.

File: pkg/ccl/changefeedccl/changefeedbase/settings.go (hunk 9 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/changefeedbase/settings.go]: Done.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed.go [hunk 12]
```diff
@@ -126,6 +126,7 @@ func Run(ctx context.Context, cfg Config) error {
 		cfg.Writer, cfg.Spans, cfg.CheckpointSpans, cfg.CheckpointTimestamp,
 		cfg.SchemaChangeEvents, cfg.SchemaChangePolicy,
 		cfg.NeedsInitialScan, cfg.WithDiff, cfg.WithFiltering,
+		changefeedbase.Quantize.Get(&cfg.Settings.SV),
```
[the remark this answers is not in the document]
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed.go:129]: Good point. Let's discuss this further offline. At first glance, having different granularities across aggregators doesn’t seem like a major issue. It might reduce quantization but may not justify the complexity of passing it to processor specs. I'll look into this tho.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 10 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:183]: Done.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 11 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go]: Done.

File: pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go (hunk 1 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:142]: Ack, I'm exploring this option now and see how much work this turns out to be. I would still prefer to drive this PR forward given this is already up, but I'm actively prototyping for this refactoring.

File: pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go [hunk 13]
```diff
@@ -128,7 +139,11 @@ func (p *rangefeed) addEventsToBuffer(ctx context.Context) error {
...
+					ev.Checkpoint.ResolvedTS = quantizeTS(ev.Checkpoint.ResolvedTS, quantize)
+				}
+				if resolvedTs := ev.Checkpoint.ResolvedTS; resolvedTs.IsEmpty() && resolvedTs.Less(p.cfg.Frontier) {
```
[the remark this answers is not in the document]
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go]: Good catch, done.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 4 above)
Comment (asg0451) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:786]: 🤷 if you think it will be helpful and it's not too spammy we can leave it in

File: pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go [hunk 14]
```diff
@@ -107,6 +109,15 @@ func (p rangefeedFactory) Run(ctx context.Context, sink kvevent.Writer, cfg rang
 	return g.Wait()
 }
 
+// quantizeTS returns a new timestamp with the walltime rounded down to the
+// nearest multiple of the quantization granularity.
+func quantizeTS(ts hlc.Timestamp, granularity time.Duration) hlc.Timestamp {
+	return hlc.Timestamp{
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:115]: nit: can you move the non-zero check into here or add one? for my peace of mind / to prevent abuse

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go [hunk 15]
```diff
@@ -645,3 +700,137 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
...
+	quantizedEntries := quantizedFrontier.Len()
+	quantizedHW := quantizedFrontier.Frontier()
+	entries := frontier.Len()
+	hw := frontier.Frontier()
+	require.LessOrEqual(t, quantizedEntries, entries)
+	if quantizedHW.WallTime == hw.WallTime {
+		require.LessOrEqual(t, quantizedHW.Logical, hw.Logical)
+	} else {
+		require.LessOrEqual(t, quantizedHW.WallTime, hw.WallTime)
+	}
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go:835]:
```suggestion
	require.True(t, quantizedHW.LessEq(hw))
```

**Review by asg0451 — APPROVED** [review]

File: pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go (hunk 14 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:119]: Done.

File: pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go (hunk 15 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go]: Done.

File: pkg/ccl/changefeedccl/changefeedbase/settings.go [hunk 16]
```diff
@@ -335,3 +335,10 @@ var DefaultLaggingRangesThreshold = 3 * time.Minute
 // DefaultLaggingRangesPollingInterval is the default polling rate at which
 // lagging ranges are checked and metrics are updated.
 var DefaultLaggingRangesPollingInterval = 1 * time.Minute
+
+var Quantize = settings.RegisterDurationSettingWithExplicitUnit(
+	settings.ApplicationLevel,
+	"changefeed.resolved_timestamp.granularity",
```
Comment (andyyang890) [inline pkg/ccl/changefeedccl/changefeedbase/settings.go:341]: Don't forget to update the commit message and PR desc with the new name.

**Review by andyyang890 — APPROVED** [review]
LGTM modulo one comment

File: pkg/ccl/changefeedccl/changefeedbase/settings.go (hunk 16 above)
Comment (wenyihu6) [reply pkg/ccl/changefeedccl/changefeedbase/settings.go:341]: Discussed this more offline - we decided to keep this cluster setting non-public and removed the release note.

**Review by wenyihu6 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138328)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @asg0451, @dt, and @rharding6373)

**Comment by wenyihu6** [comment]
TFTRs!

bors r=asg0451,andyyang890
